### ISSUE TITLE
(Auto) grab keyboard

### ISSUE THREAD

**Issue by uli42** [ArcticaProject/nx-libs#384]
https://bugs.x2go.org/cgi-bin/bugreport.cgi?bug=1011

"When in windowed mode (not full screen), and the window has the focus and
the mouse over the window, please add an option to grab the keyboard input
(for ALT-TAB within the remote session, not on the host).

As example, this is implemented in xFreeRDP and properitary Microsoft
Terminal Server Client (just as an example, i do not use Windows) and is
very very useful.

I'm using XFCE4 on Arch Linux (server and client)"

Xephyr has something like that.

**Comment by uli42** [comment]
I have done an initial implementation. Pressing ctrl-alt-g activates autograb mode. Whenever the pointer is above the nxagent window nxagent grabs all keypresses. This still has some problems, though. E.g. it grabs the keyboard even if the nxagent window is not active...

**Comment by sunweaver** [comment]
Hi Uli,

On Sun Mar 5 23:56:54 2017 GMT+0100, Ulrich Sibiller wrote:
> I have done an initial implementation. Pressing ctrl-alt-g activates autograb mode. Whenever the pointer is above the nxagent window nxagent grabs all keypresses. This still has some problems, though. E.g. it grabs the keyboard even if the nxagent window is not active... 

This is so awesome, file a PR and tag it as invalid (we need a w-i-p tag, actually) for now. This allows others to view your progress on this. 

Thanks!!!
Mike

-- 
Sent from my Fairphone (running Sailfish OS)...

**Comment by uli42** [comment]
added "invalid" to signal work-in-progress

**Comment by Ionic** [comment]
If possible, I want to backport it to X2Go's nx-libs when done.

**Comment by uli42** [comment]
Should be easy

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mihai Moldovan" <notifications@github.com>
Gesendet: ‎07.‎03.‎2017 22:46
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Assign" <assign@noreply.github.com>
Betreff: Re: [ArcticaProject/nx-libs] (Auto) grab keyboard (#384)

If possible, I want to backport it to X2Go's nx-libs when done.

**Comment by uli42** [comment]
Some (edited) chatlog regarding details of the UI:

> (09:39:11) uli42: I habe improved the autograb feature, need to cleanup the codebefore committing. However, I'd like to ask you how we design that functionality. The current state is this:
> (09:39:11) uli42: 1. run nxagent
> (09:39:11) uli42: 2. press ctrl-alt-g to toggle autograb mode.
> (09:39:11) uli42: 3. if autograb is active everytime the nxagent window gets the focus it will grab the keyboard. You then can use e.g. alt-tab to cycle through windows WITHIN the session. You can also use the mouse to leave the window and focus another window, the grab will then automatically stop. This method has one drawback: If you are cycling through windows on the native xserver and activate the nxagent window this way you cannot cycle any further until you deactivate nxagent again (legend: activate is the same as focussing here)
> (09:39:11) uli42: 3+: I have also implemented a mode where the mouse get caught within the nxagent window as soon as nxagent gets the focus. You then have to deactivate autograb by pressing ctrl-alt-g to get out again.
> (09:39:11) uli42: 4. I think I will add an option to select what mode is desired. maybe the key combination will cycle through different modes instead of simply toggling the mode.
> (09:39:11) uli42: So what do you think? How should we design that?
> (11:21:49) sunweaver: uli42: Hi!
> (11:21:55) sunweaver: awesome work!!!
> (11:22:33) sunweaver: I think that keyboard grab and mouse (pointer) grab should be handled under the same concept.
> (11:22:56) sunweaver: is there a way to enable autograb after a short delay, once the window gets active?
> (11:23:25) sunweaver: so that you can alt-tab through client-side windows, and if you don't wait to long, you can alt-tab continuously through client-side windows.
> (11:23:30) Ionic: uli42: I would probably prefer the shortcut to toggle between 3 states: normal (no grabbing), autograb and force-grab
> (11:23:55) sunweaver: what is autograb vs. force-grab?
> (11:24:25) sunweaver: then, I'd suggest that we don't have to switch off auto-grab to ungrab cursor and keyboard.
> (11:24:43) Ionic: autograb is what uli implemented, I would like to have a force-grab mode that grabs mouse and keyboard input independently of whether the window is currently active or not
> (11:24:44) sunweaver: in VirtualBox, you have to "hit" the "host-key" combination (which is configurable).
> (11:25:14) sunweaver: in NX terms, we might choose Ctrl+Alt as the release hotkey (and that should be configurable via keystrokes.cfg).
> (11:25:27) sunweaver: ah, ok.
> (11:25:37) sunweaver: so force-grab is for non-active agent window.
> (11:25:37) Ionic: ctrl-alt-g is probably better because it collides less
> (11:26:04) sunweaver: Ionic: but when releasing the pointer / keyboard, you might not want to change the grabbing mode.
> (11:26:09) sunweaver: you just want to release it.
> (11:26:12) uli42: for entering/leaving force-grab ctrl-alt-g ist too complicated, e.g. Xephyr uses ctrl-shift for that purpose
> (11:26:29) sunweaver: or that...
> (11:26:52) sunweaver: it should be a two-key combination pressable with one hand, I'd say.
> (11:26:53) Ionic: sunweaver: well, releasing essentially IS disabling autograbbing, or isn't it?
> (11:27:02) uli42: Ionic: yes, it is
> (11:27:03) sunweaver: and releasing != grab mode toggling
> (11:27:11) sunweaver: Ionic: no, it is not.
> (11:27:23) TotallyNotKim [~TotallyNo@kim-nuernberger.eu] entered the room.
> (11:27:29) sunweaver: Once you return to the session window, you'd want the same grab mode be active.
> (11:28:08) sunweaver: uli42: third thing: hitting autograb should issue an NXTransDialog call.
> (11:28:16) sunweaver: (I can add that after your PR is out).
> (11:28:29) uli42: So: We make two modes: auto-grab and force grab. Force grab needs user interaction to activate/deactivate, preferably by a convenient keystroke. Autograb will be a command line option and can be toggled while nxagent is running via ctrl-alt-g
> (11:28:34) sunweaver: so that people can potentially get some informational dialog window, when they hit ctrl+alt+g.
> (11:29:14) sunweaver: uli42: so we have
> (11:29:19) sunweaver: no grabbing vs grabbing
> (11:29:25) sunweaver: by default grabbing is auto grabbing.
> (11:29:49) sunweaver: if you then hit a convenient key stroke, you can toggle grabbing mode between auto-grabbing and forced grabbing.
> (11:29:52) sunweaver: right?
> (11:30:17) sunweaver: and grabbing mode should indeed be configurable via a command line option (and an nx/nx option).
> (11:30:19) Ionic: yes, or no grabbing at all if not activated yet
> (11:30:43) uli42: Hmm, autograb on by default? I am not sure. It changes the behaviour of nxagent comapred to the current nxagent.
> (11:31:10) Ionic: an option for that will be best, I also think it should be disabled by default
> (11:31:19) sunweaver: Ionic: I'd say that no grabbing should immitate the current state of nxagent behaviour and the grab code gets enabled by hitting ctrl+alt+g
> (11:31:31) uli42: force-grab is independent from autograb. There's no need for that to be configurable (except the keystroke) because it needs to be activated explicitely anyway.
> (11:31:32) sunweaver: when grabbing is off, it should be off really. Shouldn't it?
> (11:31:46) sunweaver: ok.
> (11:31:53) sunweaver: so we have no grab <-> force grab
> (11:31:59) sunweaver: or autograb <-> force grab
> (11:32:17) sunweaver: and with alt+ctrl+g you can switch between no grab and auto grab. right?
> (11:32:33) uli42: yes. And, as a possbile mixed-mode: auto-force-grab
> (11:32:42) sunweaver: ahhhhhhhhhh....
> (11:32:55) uli42: but I don't really like the latter
> (11:33:11) sunweaver: what would it be like in usability terms?
> (11:33:48) uli42: everytiem the nxagent window gets the focus the pointer gets locked in the window and to leave that mode the user has to press the configured force-grab key
> (11:34:05) Ionic: sunweaver: yes
> (11:34:24) Ionic: I don't like auto-force-grab either
> (11:36:06) Ionic: although uhm
> (11:37:15) Ionic: but that would make it 3 different shortcuts...
> (11:37:29) Ionic: one for force grab, one for auto grab and one for releasing
> (11:37:50) Ionic: I'd reduce it to two: force grab and auto grab, with force grab also acting as the release switch
> (11:38:57) sunweaver: can't we handle it like some other application that many people already use?
> (11:39:13) sunweaver: like VirtualBox, Xephyr (not so popular) and such?
> (11:39:32) sunweaver: so "no grab" is "no grab" (current behavious)
> (11:39:43) sunweaver: and then you can toggle to "grab"
> (11:39:54) Ionic: which is force-grab, right?
> (11:39:57) sunweaver: and when the agent is in "grab" mode, it behaves e.g. like Xephyr.
> (11:40:19) sunweaver: it depends on what other famous application we want to immitate.
> (11:41:35) Ionic: qemu grabs via clicking into the window and only releases on keypress
> (11:41:37) Ionic: for instance
> (11:41:54) Ionic: but that would be complicated in our case, because we'd have to drop the first clicking event
> (11:42:26) sunweaver: yes.
> (11:42:37) sunweaver: I quite like the Xephyr behaviour:
> (11:42:45) sunweaver: normal mode is nxagent's current mode.
> (11:42:54) sunweaver: when inside nxagent, you hit ctrl+shift.
> (11:42:59) sunweaver: and then your are "IN"
> (11:43:06) Ionic: i.e., force-grab
> (11:43:08) sunweaver: and you can only get out by Ctrl+Shift once more.
> (11:43:11) sunweaver: yes.
> (11:43:12) Ionic: and can only get out by ctrl-shift again
> (11:43:16) Ionic: i.e., force-grab :)
> (11:43:28) Ionic: and we couple that with a new auto-grab feature
> (11:43:35) sunweaver: auto-grab?
> (11:43:43) sunweaver: so what is that then?
> (11:43:47) Ionic: yes, that grabs keyboard input when nxagent has the focus
> (11:44:10) sunweaver: should that actually happen always, so that the current behaviour gets dropped fully?
> (11:44:17) Ionic: so that, if enabled (via nx/nx or parameter or shortcut), every key pres gets send to the session
> (11:44:21) sunweaver: (it has been annoying various people so far, anyway).
> (11:44:26) Ionic: no, I wouldn't make it the default
> (11:44:39) Ionic: it has been annoying various people, but others not
> (11:44:46) sunweaver: ok...
> (11:45:00) Ionic: I don't know what behavior people really expect
> (11:45:03) sunweaver: so we have grab vs. auto-grab (ALT-CTRL-g)
> (11:45:16) Ionic: but if someone doesn't have problems with the current behavior, they of course won't mention it
> (11:45:25) Ionic: because why should I mention something that works fine for me...
> (11:45:28) sunweaver: and the force-grab behaviour that Xephyr has (Shift+Ctrl to get in and out).
> (11:46:29) theUser2: can't it all be configurable?    Have one hardcoded default but let it be fine tuned for those who care
> (11:47:26) Ionic: theUser2: yes, we'll make it configurable, I'm just saying that the new auto-grab feature shouldn't be the default
> (11:49:20) sunweaver: Ionic: can you merge the misc cherries PR? I have renamed the title, but I'd say let's keep the branch name (and comment on it maybe in the merge commit msg). Preferably, we should not open a new PR, because then we loose chat backlog associated with the PR.
> (11:49:34) sunweaver: plus... on GH, branch renaming of PR branches is not possible.
> (11:49:34) theUser2: If something have been a particular way... for ages (we're talking NX so i suppose from pre jurASSic times) then yes.... the default behavior should probably be as it always was....
> (11:49:52) ***sunweaver nods about default behaviour...
> (11:50:04) theUser2: Would not want to piss of a t-rex or other predatory dino would  we?
> (11:53:32) Ionic: sunweaver: okay
> (11:53:56) sunweaver: uli42: any more confusion needed about the grabbing varieties?
> (11:56:24) uli42: no, I have enough ;-) Basically it is like I suggested, so I can go on an implement it like that. You'll be my testers, so we can iterate into another discussion once I am finished (hopefully I can do it over the weekend)

**Comment by uli42** [comment]
This is tricky. I cannot seem to find a sane way to make modifier-only key combinations like e.g. ctrl-alt or shift-ctrl configurable. I could hardcode a combination but that feels like surrender...

**Comment by uli42** [comment]
will depend on #392

**Comment by uli42** [comment]
this is still not working properly when switching to fullscreen but switch_all_screens seems to work. please test!

**Comment by leoluk** [comment]
Any progress on this?

**Comment by uli42** [comment]
Well, not really. You can try but be prepared to step over bugs. Depending
on your local window manager it can also work quite smoothly! The main
problem are intermediate switches to full screen. If you do not need them
it is really usable.

Uli

Leopold Schabel <notifications@github.com> schrieb am Do., 14. Juni 2018,
19:11:

**Comment by daliu87** [comment]
Any updates on this feature?  Would be great to have, using awesomewm server on a windows client.

Thanks!

**Comment by uli42** [comment]
https://github.com/uli42/nx-libs/tree/pr/autograb contains my version. You might want to try that version. But beware. It is not perfect and the code is 2 years old and ~1000 commits behind the current version.

It largely depends on your window manager if it is working smoothly or not. And you should avoid switching to fullscreen via keystroke. There autograb and inputlock. autograb (ctrl-alt-g) means "grab the keyboard when the NX window gets focus". inputlock (ctrl-alt-c) means "lock the pointer in the input window"

**Comment by qpkorr** [comment]
Not sure how close this is to being released, but just wanted to say thank you for your continued work on it - I too am on a windows client and an awesomewm server, and would love to have alt-tab available.
